readers/mutation_reader.cc

```diff
@@ -39,111 +41,149 @@ mutation_fragment_stream_validator::mutation_fragment_stream_validator(const ::s
-    return (*this)(to_mutation_fragment_kind_v2(mf.mutation_fragment_kind()), mf.position());
-}
+    auto valid = true;
 
-bool mutation_fragment_stream_validator::operator()(mutation_fragment_v2::kind kind) {
-    bool valid = true;
+    // Check fragment kind order
     switch (_prev_kind) {
         case mutation_fragment_v2::kind::partition_start:
-            valid = kind != mutation_fragment_v2::kind::partition_start;
+            valid &= kind != mutation_fragment_v2::kind::partition_start;
             break;
         case mutation_fragment_v2::kind::static_row: // fall-through
         case mutation_fragment_v2::kind::clustering_row: // fall-through
         case mutation_fragment_v2::kind::range_tombstone_change:
-            valid = kind != mutation_fragment_v2::kind::partition_start &&
+            valid &= kind != mutation_fragment_v2::kind::partition_start &&
                     kind != mutation_fragment_v2::kind::static_row;
             break;
         case mutation_fragment_v2::kind::partition_end:
-            valid = kind == mutation_fragment_v2::kind::partition_start;
+            valid &= kind == mutation_fragment_v2::kind::partition_start;
             break;
     }
-    if (valid) {
-        _prev_kind = kind;
+    if (!valid) {
+        return false;
+    }
+
+    if (pos && _prev_kind != mutation_fragment_v2::kind::partition_end) {
+        auto cmp = position_in_partition::tri_compare(_schema);
+        auto res = cmp(_prev_pos, *pos);
+        if (_prev_kind == mutation_fragment_v2::kind::range_tombstone_change) {
+            valid &= res <= 0;
+        } else {
+            valid &= res < 0;
+        }
+        if (!valid) {
+            return false;
+        }
+    }
+
+    _prev_kind = kind;
+    if (pos) {
+        _prev_pos = *pos;
+    } else {
+        switch (kind) {
+            case mutation_fragment_v2::kind::partition_start:
+                _prev_pos = position_in_partition(position_in_partition::partition_start_tag_t{});
+                break;
+            case mutation_fragment_v2::kind::static_row:
+                _prev_pos = position_in_partition(position_in_partition::static_row_tag_t{});
+                break;
+            case mutation_fragment_v2::kind::clustering_row:
+            case mutation_fragment_v2::kind::range_tombstone_change: // fall-through
+                if (!_prev_pos.has_clustering_key()) { // don't move pos if it is already a clustering one
+                    _prev_pos = position_in_partition(position_in_partition::before_clustering_row_tag_t{}, clustering_key::make_empty());
+                }
+                break;
+            case mutation_fragment_v2::kind::partition_end:
+                _prev_pos = position_in_partition(position_in_partition::end_of_partition_tag_t{});
+                break;
+        }
     }
-    return valid;
+    _current_tombstone = new_current_tombstone.value_or(_current_tombstone);
```

Comment:
why not prevent writing to memory if new_current_tombstone is disengaged?
this is on the hot path and value_or doesn't save the conditional anyhow...

Reply:
I was optimizing for number of source lines. I agree an if here makes more sense.